### DISCUSSION TITLE
Issue controling threads from MQTT

### DISCUSSION THREAD

**Discussion by Pilou44** [micropython/micropython#16426] [General]
Hi,

I'm trying to control pixel led strip from MQTT. It works fine as long as it's just switch on and off, but if I use thread to animate leds I just can't stop it. My full code is available here : https://github.com/Pilou44/mqttSwitch/blob/main/pico/pico.py

Here is my loop for MQTT reading messages : 
```
def run():
    global client
    # Initialize Wi-Fi
    if not initialize_wifi(config.wifi_ssid, config.wifi_password):
        print('Error connecting to the network... exiting program')
    else:
        # Connect to MQTT broker, start MQTT client
        client = connect_mqtt()
        publish_discovery()
        client.set_callback(my_callback)
        
        client.subscribe(CMD_TOPIC)
        
        # Continuously checking for messages
        while True:
            sleep(1)
            client.check_msg()
            print('Loop running')
```
And my LED control functions:
```
def switch_on():
    enabled = True
    pixels[0] = (255, 0, 0) # set to red, full brightness
    pixels[1] = (0, 128, 0) # set to green, half brightness
    pixels[2] = (0, 0, 64)  # set to blue, quarter brightness
    pixels.write()
    print("Launch thread")
    _thread.start_new_thread(thread_loop, ())
        
def thread_loop():
    print("Thread started")
    global enabled
    print(f"Emabled = {enabled}")
    enabled = True
    print(f"Emabled = {enabled}")
    while (enabled):
        sleep_ms(500)
        rotate(pixels)
    print("Thread ended")

def switch_off():
    global enabled
    print("Switch off")
    enabled = False
    print(f"Emabled = {enabled}")
    pixels[0] = (0, 0, 0)
    pixels[1] = (0, 0, 0)
    pixels[2] = (0, 0, 0)
    pixels.write()
```

On executing, command to switch on works, and I keep seeing message `Loop running` from `run` function. But no other MQTT messages are received. If I don't start the `thread_loop` then I can start and stop leds as long as I want.

Does anyone see what the problem is ?
Thank you

**Comment by peterhinch** [comment]
Sharing data between threads is difficult - see [this doc](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md). The solution which avoids tackling that complexity is to use asyncio - see [docs](https://docs.micropython.org/en/latest/library/asyncio.html) and [tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md).

**Reply by Pilou44** [reply]
Thank you
